### DISCUSSION TITLE
SQLAlchemy dynamic schema-names does not work if session is returned from try/except or with-block

### DISCUSSION THREAD

**Discussion by alispa** [sqlalchemy/sqlalchemy#8629] [Usage Questions]
Hello.

I am trying to use multiple schemas in SQLAlchemy. Somehow it does not work properly when using context manager (`with-block`) or `try/except-block`.

This is my short execution example:

```
def import_units(schema_name: str):
    from taxes.models import Tax
    
    dbs = get_db_session(schema_name)
    dbs.add(Tax(percentage=19.00, abbreviation="abc"))
    dbs.commit() ## during commit I get ERROR:
       ## sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedTable) relation "per_user.taxes" does not exist
       ## LINE 1: INSERT INTO per_user.taxes (percentage, abbreviation, de...
```

Here is how I generate the `session `and pass it with and without `blocks`:

```
def get_db_session(schema_name: str):    
    dbs = SessionLocal()
    dbs.connection(execution_options={"schema_translate_map": {'per_user': schema_name}})
    
    return dbs ## this return WORKS (translates into right schema name)

    try:
        return dbs ## this does NOT work (no translation, uses 'per_user' as schema name)
    finally:
        dbs.close()

    with SessionLocal() as s:
        s.connection(execution_options={"schema_translate_map": {'per_user': schema_name}})
        return s ## this does NOT work (no translation, uses 'per_user' as schema name)
```

If the `imports `and `session `creation can give you some data, here are they too.

```
from config.settings import SQL_DATABASE_URI
import warnings
from sqlalchemy import create_engine, exc
from sqlalchemy.orm import sessionmaker

engine = create_engine(SQL_DATABASE_URI, pool_pre_ping=True, future=True)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine, future=True)
warnings.filterwarnings("always", category=exc.RemovedIn20Warning)
```

**Comment by zzzeek** [comment]
when you .close() the Session within your try/finally block, the Session object is closed before it's returned.  So your schema map assignment is lost.  The same thing will occur in your "with" block; "return" is not actually evaluated until those blocks are complete.

A way to do the pattern you are looking for would still require that the caller notes where the block should end; however once you set that up, you can encapsulate your session construction as:

```py

import contextlib

@contextlib.contextmanager
def get_db_session(schema_name: str):    
    dbs = SessionLocal()
    dbs.connection(execution_options={"schema_translate_map": {'per_user': schema_name}})
    
    try:
        yield dbs
    finally:
        dbs.close()

```

calling code would then work as:

```py
with get_db_session("schema") as session:
    session.add(...)
    session.commit()
```


see [@contextlib.contextmanager](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager)

**Comment by alispa** [comment]
Thank you for your answer. It works.

Is this solution applicable for the `FastAPI `approach for `session dependencies`?

This is their example from the [docs](https://fastapi.tiangolo.com/tutorial/sql-databases/?h=session#create-a-dependency):
```
# Dependency
def get_db():
    db = SessionLocal()
    try:
        yield db
    finally:
        db.close()
```

```
@app.post("/users/", response_model=schemas.User)
def create_user(user: schemas.UserCreate, db: Session = Depends(get_db)):
    return crud.get_user_by_email(db, email=user.email)
```

I've implemented your solution and tried to integrate it into the `FastAPI` example, like:

```
def get_db_session_dependency(req: Request):
    schema_name = req.headers['schema_name'] # pass schema name in headers
    with get_db_session(schema_name) as session:
        yield session
```

```
@app.post("/users/", response_model=schemas.User)
def create_user(user: schemas.UserCreate, db: Session = Depends(get_db_session_dependency)):
    return crud.get_user_by_email(db, email=user.email)
```

```
def get_user_by_email(db_session: Session, email: str):
    u = db_session.execute(select(User).filter_by(email=email)).scalar_one_or_none()
    db_session.commit()
    return u
```

But during `.commit()` the schema still remains 'per_user' (untranslated).

**Reply by zzzeek** [reply]
> Thank you for your answer. It works.
> 
> Is this solution applicable for the `FastAPI `approach for `session dependencies`?
> 
> This is their example from the [docs](https://fastapi.tiangolo.com/tutorial/sql-databases/?h=session#create-a-dependency):
> 
> ```
> # Dependency
> def get_db():
>     db = SessionLocal()
>     try:
>         yield db
>     finally:
>         db.close()
> ```
> 
> ```
> @app.post("/users/", response_model=schemas.User)
> def create_user(user: schemas.UserCreate, db: Session = Depends(get_db)):
>     return crud.get_user_by_email(db, email=user.email)
> ```
> 

That looks like they are doing something similar, `Depends()` I assume holds onto the iterable, and FastAPI knows how to close out the iterable at the end so that it reaches the "finally:".  So assuming that's what that is, that would work also.


> I've implemented your solution and tried to integrate it into the `FastAPI` example, like:
> 
> ```
> def get_db_session_dependency(req: Request):
>     schema_name = req.headers['schema_name'] # pass schema name in headers
>     with get_db_session(schema_name) as session:
>         yield session
> ```

OK well is that context manager also being integrated with the `Depends` thing?  i dont know that those interoperate.

> 
> ```
> @app.post("/users/", response_model=schemas.User)
> def create_user(user: schemas.UserCreate, db: Session = Depends(get_db_session_dependency)):
>     return crud.get_user_by_email(db, email=user.email)
> ```

right I dont know that that will work.   Maybe just use the Depends by itself.   the main thing is, don't use "return" in the session function that is going to be closing inside of a finally: block.

> 
> ```
> def get_user_by_email(db_session: Session, email: str):
>     u = db_session.execute(select(User).filter_by(email=email)).scalar_one_or_none()
>     db_session.commit()
>     return u
> ```
> 
> But during `.commit()` the schema still remains 'per_user' (untranslated).

look in your SQL logs and make sure you don't see a second transaction, you would see BEGIN..COMMIT.  if you see BEGIN..ROLLBACK..BEGIN..COMMIT that means your session is getting closed.  Or just put a print statement in that finally() to see when it happens.

**Comment by alispa** [comment]
Thanks. I will do some debugging and report the results.

An update, to document the situation... 
Just adding `execution_options` to the FastAPI [example code](https://fastapi.tiangolo.com/tutorial/sql-databases/?h=session#create-a-dependency) does not work (schema name remains 'per_user'):
```
def get_db_session(req: Request):
    schema_name = req.headers['schema_name'] # pass schema name in headers
    db_session = SessionLocal()
    db_session.connection(
        execution_options={
            "schema_translate_map": {'per_user': schema_name}
        }
    )
    try:
        yield db_session
    finally:
        db_session.close()
```
It is called like this:
```
@app.post("/")
def test(db: Session = Depends(get_db_session)):
    db.execute(something)
    db.commit()
```

**Reply by CaselIT** [reply]
I think you would get better support by asking in the fastapi support page. I, like mike, don't know much about fast api, sorry

**Reply by alispa** [reply]
Thank you guys for your effort.
It seems that I found the problem, but I have to test in detail.

I have a pretty big app, it worked properly without multiple schemas.
Now, with multiple schemas, it seems that the problem appears when you pass the session into a function and in that function call .commit().
Then, if you pass the same session into another function the dynamic schema name gets lost.
I applied .flush() instead of .commit() and for not nothing breaks.

**Reply by zzzeek** [reply]
OK this is all expected because .connection() only lasts until the next commit.

I dont know what SessionLocal is.  but if you make a Session, you can weld it to a certain schema by setting the options on the engine instead:

```py
sess = Session(bind=engine.execution_options({"schema_translate_map": {'per_user': schema_name}})
```

the above map will stick on that Session for its lifespan.
